svm-rent-collector/src/svm_rent_collector.rs

```diff
@@ -0,0 +1,137 @@
+//! Plugin trait for rent collection within the Solana SVM.
+
+use {
+    crate::rent_state::RentState,
+    solana_sdk::{
+        account::{AccountSharedData, ReadableAccount},
+        clock::Epoch,
+        pubkey::Pubkey,
+        rent::{Rent, RentDue},
+        rent_collector::CollectedInfo,
+        transaction::{Result, TransactionError},
+        transaction_context::{IndexOfAccount, TransactionContext},
+    },
+};
+
+mod rent_collector;
+
+/// Rent collector trait. Represents an entity that can evaluate the rent state
+/// of an account, determine rent due, and collect rent.
+///
+/// Implementors are responsible for evaluating rent due and collecting rent
+/// from accounts, if required. Methods for evaluating account rent state have
+/// default implementations, which can be overridden for customized rent
+/// management.
+pub trait SVMRentCollector {
+    /// Check rent state transition for an account in a transaction.
+    ///
+    /// This method has a default implementation that calls into
+    /// `check_rent_state_with_account`.
+    fn check_rent_state(
+        &self,
+        pre_rent_state: Option<&RentState>,
+        post_rent_state: Option<&RentState>,
+        transaction_context: &TransactionContext,
+        index: IndexOfAccount,
+    ) -> Result<()> {
+        if let Some((pre_rent_state, post_rent_state)) = pre_rent_state.zip(post_rent_state) {
+            let expect_msg =
+                "account must exist at TransactionContext index if rent-states are Some";
+            self.check_rent_state_with_account(
+                pre_rent_state,
+                post_rent_state,
+                transaction_context
+                    .get_key_of_account_at_index(index)
+                    .expect(expect_msg),
+                &transaction_context
+                    .get_account_at_index(index)
+                    .expect(expect_msg)
+                    .borrow(),
+                index,
+            )?;
+        }
+        Ok(())
+    }
+
+    /// Check rent state transition for an account directly.
+    ///
+    /// This method has a default implementation that checks whether the
+    /// transition is allowed and returns an error if it is not. It also
+    /// verifies that the account is not the incinerator.
+    fn check_rent_state_with_account(
+        &self,
+        pre_rent_state: &RentState,
+        post_rent_state: &RentState,
+        address: &Pubkey,
+        _account_state: &AccountSharedData,
+        account_index: IndexOfAccount,
+    ) -> Result<()> {
+        if !solana_sdk::incinerator::check_id(address)
+            && !self.transition_allowed(pre_rent_state, post_rent_state)
+        {
+            let account_index = account_index as u8;
+            Err(TransactionError::InsufficientFundsForRent { account_index })
+        } else {
+            Ok(())
+        }
+    }
```

Comment:
Here I've used the same implementation as the original, but I've omitted the `debug!` to avoid a dependency on `log` and the metrics submission. The former omission is why `_account_state` is unused here.

Reply:
As mentioned in the PR description, Bank's wrapper around `RentCollector` would then override this method to inject `debug!` and metrics submission.